contracts/voting-verifier/src/execute.rs

```diff
@@ -344,3 +346,174 @@ fn remove_pending_message(
 
     Ok(pending_messages)
 }
+
+fn cast_vote(
+    store: &mut dyn Storage,
+    block_height: u64,
+    poll: &mut WeightedPoll,
+    voter: &Addr,
+    votes: Vec<bool>,
+) -> Result<(), ContractError> {
+    VOTED.update(
+        store,
+        (poll.poll_id, voter.clone()),
+        |v| -> Result<(), ContractError> {
+            match v {
+                Some(_) => Err(ContractError::AlreadyVoted),
+                None => {
+                    poll.cast_vote(block_height, voter, votes)?;
+                    Ok(())
+                }
+            }
+        },
+    )
+}
+
+fn poll_result(
+    store: &mut dyn Storage,
+    block_height: u64,
+    poll: &mut WeightedPoll,
+) -> Result<PollResult, ContractError> {
+    let voters = VOTED
+        .prefix(poll.poll_id)
+        .range(store, None, None, Order::Ascending)
+        .count();
+
+    if block_height < poll.expires_at
+            // can tally early if all participants voted
+            && voters < poll.snapshot.get_participants().len()
+    {
+        return Err(ContractError::PollNotEnded);
+    }
+
+    Ok(poll.tally()?)
```

Comment:
if this construction is just about matching error types I'd prefer `poll.tally().into()`